authentik/sources/kerberos/views.py

```diff
@@ -0,0 +1,171 @@
+        """Get GSSAPI server context from cache or create it"""
+        server_creds = self.source.get_gssapi_creds()
+        if server_creds is None:
+            return None
+
+        state = cache.get(f"{SPNEGO_STATE_CACHE_PREFIX}/{key}", None)
+
+        if state:
+            # pylint: disable=c-extension-no-member
+            return gssapi.sec_contexts.SecurityContext(
+                base=gssapi.raw.sec_contexts.import_sec_context(state),
+            )
+
+        return gssapi.sec_contexts.SecurityContext(creds=server_creds, usage="accept")
+
+    def set_server_ctx(self, key: str, ctx: gssapi.sec_contexts.SecurityContext):
+        """Store the GSSAPI server context in cache"""
+        cache.set(f"{SPNEGO_STATE_CACHE_PREFIX}/{key}", ctx.export(), SPNEGO_STATE_CACHE_TIMEOUT)
+
+    # pylint: disable=too-many-return-statements
+    def dispatch(self, request, *args, **kwargs) -> HttpResponse:
+        """Process SPNEGO request"""
+        self.source: KerberosSource = get_object_or_404(
+            KerberosSource,
+            slug=kwargs.get("source_slug", ""),
+            enabled=True,
+        )
+
+        qstring = request.GET if request.method == "GET" else request.POST
+        state = qstring.get("state", None)
+        if not state:
+            return redirect(
+                reverse(
+                    "authentik_sources_kerberos:spnego-login",
+                    kwargs={"source_slug": self.source.slug},
+                )
+                + f"?state={self.new_state()}"
+            )
+
+        authstr = self.get_authstr(request)
+        if not authstr:
+            LOGGER.debug("authstr not present, sending challenge")
+            return self.challenge(request)
+
+        try:
+            in_token = b64decode(authstr)
+        except (TypeError, ValueError):
+            return self.challenge(request)
+
+        with Krb5ConfContext(self.source):
+            server_ctx = self.get_server_ctx(state)
+            if not server_ctx:
+                return self.challenge(request)
+
+            try:
+                out_token = server_ctx.step(in_token)
+            except gssapi.exceptions.GSSError as exc:
+                LOGGER.debug("GSSAPI security context failure", exc=exc)
+                return self.challenge(request)
+
+            if not server_ctx.complete or server_ctx.initiator_name is None:
+                self.set_server_ctx(state, server_ctx)
+                return self.challenge(request, out_token)
+
+            identifier = str(server_ctx.initiator_name)
+            context = {
+                "spnego_info": {
+                    "initiator_name": str(server_ctx.initiator_name),
+                    "target_name": str(server_ctx.target_name),
+                    "mech": str(server_ctx.mech),
+                    "actual_flags": server_ctx.actual_flags,
+                },
+            }
+
+        return SPNEGOSourceFlowManager(
```

Comment:
You are supposed to send the `out_token` in `WWW-AUTHENTICATE` header to the client for mutual authentication purposes.